src/main/java/frc/robot/subsystems/TurretSubsystem.java

```diff
@@ -0,0 +1,109 @@
+// Copyright (c) FIRST and other WPILib contributors.
+// Open Source Software; you can modify and/or share it under the terms of
+// the WPILib BSD license file in the root directory of this project.
+
+package frc.robot.subsystems;
+
+import com.ctre.phoenix6.hardware.CANcoder;
+import com.revrobotics.CANSparkLowLevel.MotorType;
+import com.revrobotics.CANSparkMax;
+import edu.wpi.first.math.controller.PIDController;
+import edu.wpi.first.math.geometry.Pose2d;
+import edu.wpi.first.wpilibj2.command.Command;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.Constants;
+import frc.robot.LimelightHelpers;
+import frc.robot.networktables.DebugEntry;
+
+public class TurretSubsystem extends SubsystemBase {
+
+  private CANSparkMax turretMotor;
+  private PIDController turretPIDController;
+  private CANcoder m_encoder;
+  private double turretPosition;
+  private double turretVelocity;
+  private DebugEntry<Double> turretPositionEntry =
+      new DebugEntry<Double>(turretPosition, "Position", this);
+  private DebugEntry<Double> turretVelocityEntry =
+      new DebugEntry<Double>(turretVelocity, "Velocity", this);
+
+  public TurretSubsystem() {
+    turretMotor = new CANSparkMax(Constants.TurretConstants.TURRET_MOTOR_ID, MotorType.kBrushless);
+    turretMotor.restoreFactoryDefaults();
+    turretMotor.setSmartCurrentLimit(40);
+
+    turretMotor.enableSoftLimit(CANSparkMax.SoftLimitDirection.kForward, true);
+    turretMotor.enableSoftLimit(CANSparkMax.SoftLimitDirection.kReverse, true);
+
+    turretMotor.setSoftLimit(CANSparkMax.SoftLimitDirection.kForward, 1);
+    turretMotor.setSoftLimit(CANSparkMax.SoftLimitDirection.kReverse, -1);
+
+    // initialze PID controller and encoder objects
+    turretPIDController =
+        new PIDController(
+            Constants.TurretConstants.TURRET_KP,
+            Constants.TurretConstants.TURRET_KI,
+            Constants.TurretConstants.TURRET_KD);
+    m_encoder = new CANcoder(Constants.TurretConstants.CANcoder_ID);
+
+    // set PID coefficients
+    turretPIDController.setP(Constants.TurretConstants.TURRET_KP);
+    turretPIDController.setI(Constants.TurretConstants.TURRET_KI);
+    turretPIDController.setD(Constants.TurretConstants.TURRET_KD);
+    turretPIDController.setIZone(Constants.TurretConstants.TURRET_KIZ);
+  }
+
+  public void stopTurret() {
+    turretMotor.stopMotor();
+  }
+
+  public void setTurretPos(double setpoint) {
+    turretMotor.set(turretPIDController.calculate(turretPosition, setpoint));
```

Comment:
Additionally, we will want to always be sending updated values to the motor so that it holds it's position even when we aren't actively commanding it. So it will react to a defender collision causing it to shift, for instance. Consider either setting up a default command that holds position or always calling turretMotor.set() in periodic.